packages/core/src/components/CheckboxTree/CheckboxTree.tsx

```diff
@@ -0,0 +1,303 @@
+import {
+  List,
+  ListItem,
+  ListItemIcon,
+  Checkbox,
+  ListItemText,
+  Collapse,
+  Typography,
+} from '@material-ui/core';
+import ExpandLess from '@material-ui/icons/ExpandLess';
+import ExpandMore from '@material-ui/icons/ExpandMore';
+
+type IndexedObject<T> = {
+  [key: string]: T;
+};
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    root: {
+      width: '100%',
+      minWidth: 10,
+      maxWidth: 360,
+      backgroundColor: 'transparent',
+      '&:hover': {
+        backgroundColor: 'transparent',
+      },
+      '&:active': {
+        animation: 'none',
+        transform: 'none',
+      },
+    },
+    nested: {
+      paddingLeft: theme.spacing(5),
+      height: '32px',
+      '&:hover': {
+        backgroundColor: 'transparent',
+      },
+    },
+    listItemIcon: {
+      minWidth: 10,
+    },
+    listItem: {
+      '&:hover': {
+        backgroundColor: 'transparent',
+      },
+    },
+    text: {
+      '& span, & svg': {
+        fontWeight: 'normal',
+        fontSize: 14,
+      },
+    },
+  }),
+);
+
+/* SUB_CATEGORY */
+
+type SubCategory = {
+  label: string;
+  isChecked?: boolean;
+  isOpen?: boolean;
+  options?: Option[];
+};
+
+type SubCategoryWithIndexedOptions = {
+  label: string;
+  isChecked?: boolean;
+  isOpen?: boolean;
+  options: IndexedObject<Option>;
+};
+
+/* OPTION */
+
+type Option = {
+  label: string;
+  value: string | number;
+  isChecked?: boolean;
+};
+
+export type CheckboxTreeProps = {
+  subCategories: SubCategory[];
+  label: string;
+  onChange: (arg: any) => any;
+};
+
+/* REDUCER */
+
+type checkOptionPayload = {
+  subCategoryLabel: string;
+  optionLabel: string;
+};
+
+type Action =
+  | { type: 'checkOption'; payload: checkOptionPayload }
+  | { type: 'checkCategory'; payload: string }
+  | { type: 'openCategory'; payload: string };
+
+const checkAllOptions = (
+  arr: Option[],
+  isChecked: boolean,
+): IndexedObject<Option> =>
+  arr.reduce((accumulator, el) => {
+    return {
+      ...accumulator,
+      [el.label]: { ...el, isChecked },
+    };
+  }, {});
+
+const reducer = (
+  state: IndexedObject<SubCategoryWithIndexedOptions>,
+  action: Action,
+) => {
+  switch (action.type) {
+    case 'checkOption': {
+      const newOptions = {
+        ...state[action.payload.subCategoryLabel].options,
+        [action.payload.optionLabel]: {
+          ...state[action.payload.subCategoryLabel].options[
+            action.payload.optionLabel
+          ],
+          isChecked: !state[action.payload.subCategoryLabel].options[
+            action.payload.optionLabel
+          ].isChecked,
+        },
+      };
+
+      return {
+        ...state,
+        [action.payload.subCategoryLabel]: {
+          ...state[action.payload.subCategoryLabel],
+          isChecked: Object.values(newOptions).every(
+            option => option.isChecked,
+          ),
+          options: newOptions,
+        },
+      };
+    }
```

Comment:
Reducers in plain js tend to get ultra gnarly. Would something like immer help you condense these into much more readable code? All of the spreads will go away.

https://immerjs.github.io/immer/docs/introduction

```ts
return produce(state, newState => {
  const category = newState[action.payload.subCategoryLabel];
  const option = category.options[action.payload.optionLabel];
  option.selected = !option.selected;
  category.selected = category.options.every(o => o.selected);
});
```